File: output/xCodeEval_8000/C++/Python/2ba8bd30f8f86980a15291e0ca997648.py
import sys

def dijkstra(graph, start):
    n = len(graph)
    dist = [float('inf')] * n
    dist[start] = 0
    visited = [False] * n

    for _ in range(n):
        u = -1
        for i in range(n):
            if not visited[i] and (u == -1 or dist[i] < dist[u]):
                u = i
        visited[u] = True

        for v, w in graph[u]:
            if dist[u] + w < dist[v]:
                dist[v] = dist[u] + w

    return dist

def solve(n, k, arrs):
    sums = [sum(arr) for arr in arrs]
    dp = [[0] * (k + 1) for _ in range(n + 1)]

    for i in range(1, n + 1):
        dp[i] = dp[i - 1].copy()
        for j in range(k, len(arrs[i - 1]) - 1, -1):
            dp[i][j] = max(dp[i][j], dp[i][j - len(arrs[i - 1])] + sums[i - 1])

    for i in range(1, n + 1):
        for j in range(k + 1):
            for l in range(min(k, len(arrs[i - 1])) + 1):
                if j - l >= 0:
                    dp[i][j] = max(dp[i][j], dp[i - 1][j - l] + sum(arrs[i - 1][:l]))

    return dp[n][k]

def main():
    n, k = map(int, input().split())
    arrs = []
    for _ in range(n):
        t = int(input())
        arr = list(map(int, input().split()))
        if t > k:
            arr = arr[:k]
        arrs.append(arr)
    print(solve(n, k, arrs))

if __name__ == "__main__":
    main()

File: output/xCodeEval_8000/C++/Go/2ba8bd30f8f86980a15291e0ca997648.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"sort"
	"strconv"
)

type Solver struct {
	in  *bufio.Reader
	out io.Writer
}

func NewSolver(r io.Reader, w io.Writer) *Solver {
	return &Solver{in: bufio.NewReader(r), out: w}
}

func (s *Solver) readInt() int {
	token, _ := s.in.ReadBytes('\n')
	num, _ := strconv.Atoi(string(token[:len(token)-1]))
	return num
}

func (s *Solver) readInts(n int) []int {
	ints := make([]int, n)
	for i := 0; i < n; i++ {
		ints[i] = s.readInt()
	}
	return ints
}

func (s *Solver) sum(arr []int) int {
	total := 0
	for _, num := range arr {
		total += num
	}
	return total
}

func (s *Solver) solve() {
	n := s.readInt()
	k := s.readInt()
	sums := make([]int, n)
	arrs := make([][]int, n)
	for i := 0; i < n; i++ {
		t := s.readInt()
		arrs[i] = s.readInts(t)
		if t > k {
			arrs[i] = arrs[i][:k]
		}
		sums[i] = s.sum(arrs[i])
	}

	dps := make([][]int, 13)
	for i := 0; i < 13; i++ {
		dps[i] = make([]int, k+1)
	}

	var rec func(l, r, dpi int) int
	rec = func(l, r, dpi int) int {
		ans := 0
		carr := arrs[l]
		dp := dps[dpi]
		if l+1 == r {
			if dp[k] > ans {
				ans = dp[k]
			}
			sum := 0
			for i := 0; i < min(k, len(carr)); i++ {
				sum += carr[i]
				if dp[k-i-1]+sum > ans {
					ans = dp[k-i-1] + sum
				}
			}
		} else {
			m := (l + r) / 2
			copy(dps[dpi+1], dps[dpi])
			for i := l; i < m; i++ {
				v := sums[i]
				k1 := len(arrs[i])
				for j := k; j >= k1; j-- {
					if dps[dpi+1][j] < dps[dpi][j-k1]+v {
						dps[dpi+1][j] = dps[dpi][j-k1] + v
					}
				}
			}
			if res := rec(m, r, dpi+1); res > ans {
				ans = res
			}
			copy(dps[dpi+1], dps[dpi])
			for i := m; i < r; i++ {
				v := sums[i]
				k1 := len(arrs[i])
				for j := k; j >= k1; j-- {
					if dps[dpi+1][j] < dps[dpi][j-k1]+v {
						dps[dpi+1][j] = dps[dpi][j-k1] + v
					}
				}
			}
			if res := rec(l, m, dpi+1); res > ans {
				ans = res
			}
		}
		return ans
	}

	fmt.Fprintln(s.out, rec(0, n, 0))
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func main() {
	solver := NewSolver(os.Stdin, os.Stdout)
	solver.solve()
}
